### PR TITLE
New Github Workflows

### PR DESCRIPTION
**Pull request by dc-andysign**
---
name: New Github Workflows
about: Changed workflow yaml files for test and release
---

## Summary

This PR introduces two new files, more exactly **`.github/workflows/test.yml`** and **`.github/workflows/release.yml`**.

The first one contains 6 jobs ( grab repo ids, lint, dependency check, tests, check package and package-lock versions and build docker ).

The second one also contains 6 jobs ( preconditions, lint, dependency check, tests, check package and package-lock versions and docker publish ).

---

## Motivation

To make this project it easier to maintain the two files in github workflows, as in test and release need to follow DC's way of doing workflows.

---

## Describe alternatives you've considered

An alternative explored was to a solution closer to the old set of Github Workflow files.

---

## Additional context

**Additional items to be discussed**:

To save resources when a new commit appears the previous _in-progress_ job gets canceled.

```yaml
concurrency:
  group: aries-framework-${{ github.ref }}-${{ github.repository }}-${{ github.event_name }}
  cancel-in-progress: true
```

Concurrency rules has been added, aka, according to google, _To cancel any currently running job or workflow in the same concurrency group, specify: cancel-in-progress: true_.

---

In order suppress a couple of errors the following depcheck config file ( **`.depcheckrc`** ) has been added:

```json
{
  "ignores": [
    "ref-struct-di",
    "@types/multer",
    "@types/ref-struct-di",
    "tsconfig-paths"
  ]
}
```

---

Added skip to **`connection.test.ts`**  , **`ConnectionController`** > **`Connection WebSocket Event`**

```js
  describe('Connection WebSocket Event', () => {
    test.skip('should return connection event sent from test agent to websocket client', async () => {
      /* ... */
    })
  })
```

Also in **`webhook.test.ts`** added _sleep_ in _after () => {}_

```js
  after(async () => {
    await new Promise(r => setTimeout(r, 2000))
    await aliceAgent.shutdown()
    await aliceAgent.wallet.delete()
    await bobAgent.shutdown()
    await bobAgent.wallet.delete()
    server.close()
  })
```

This is designed to suppress the unit test race condition errors. This is important because otherwise the job after the **`test`** job in **`test.yml`** will not get executed.

---

It has been suggested that both platforms ( _amd64_ and _arm64_ ) should be added to both **`test.yml`** and **`release.yml`** but, adding platforms **`linux/arm64`** makes workflow extremely slow, like one test took **54m 58s** ( See [https://github.com/digicatapult/veritable-cloudagent/actions/runs/5861569817](https://github.com/digicatapult/veritable-cloudagent/actions/runs/5861569817) ). In that respect, the platform **`linux/arm64`** will only be added to the **`release.yml`** file.

E.g. extract from the **`test.yml`** workflows file:

```yml
  build-docker:
    name: "Build docker image"
    runs-on: ubuntu-latest
    steps:
      # ...
      - name: Build image
        uses: docker/build-push-action@v4
        with:
          builder: ${{ steps.buildx.outputs.name }}
          context: .
          file: ./afj-arm.dockerfile
          platforms: linux/amd64
          push: false
```

---

Note that there is an issue and when one creates a new release, the string that represents the version number will contain quotes:

This is due to the fact that the variable VERSION needs to be sanitized:

```
VERSION: v"0.2.0"
```

See the releases page:

**[https://github.com/digicatapult/veritable-cloudagent/releases](https://github.com/digicatapult/veritable-cloudagent/releases)**

Example of releases that have this problem in the name: **`Release v"0.0.1"`**, **`Latest Release v"0.0.1"`**, **`Release v"0.1.2"`**

That being said this is not an issue as the release pipeline still works despite the existence of this double quote character.

---

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -76,94 +122,102 @@ jobs:
...
+          username: ${{ secrets.DSCP_DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DSCP_DOCKERHUB_TOKEN }}
+      - name: Build image
+        uses: docker/build-push-action@v4
+        with:
+          builder: ${{ steps.buildx.outputs.name }}
+          context: .
+          file: ./afj-arm.dockerfile
+          platforms: linux/amd64, linux/arm64
+          # platforms: linux/amd64
```
Comment (jonmattgray) [inline .github/workflows/release.yml:190]: delete?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -19,130 +14,132 @@ concurrency:
...
+        run: npm install # maybe this should be changed to npm ci
+      - name: Dependency Check
+        run: npm run depcheck
 
-  package-finder:
+  tests:
+    name: Run tests
+    strategy:
+      matrix:
+        test-type: [unit]
```
Comment (jonmattgray) [inline .github/workflows/test.yml:84]: this as well can maybe be deleted

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -19,130 +14,132 @@ concurrency:
...
-          node-version: 18
-        if: ${{ steps.release-please.outputs.release_created }}
-      - name: Install dependencies
-        run: npm ci
-        if: ${{ steps.release-please.outputs.release_created }}
+          builder: ${{ steps.buildx.outputs.name }}
+          context: .
+          file: ./afj-arm.dockerfile
+          # platforms: linux/amd64, linux/arm64
+          platforms: linux/amd64
```
Comment (jonmattgray) [inline .github/workflows/test.yml:144]: change back to `platforms: linux/amd64, linux/arm64`?

Comment (jonmattgray) [reply .github/workflows/test.yml:144]: sorry, read the PR description, maybe add a comment here about why `arm64` isn't being built?

Comment (mattdean-digicatapult) [reply .github/workflows/test.yml:144]: and remove the commented out line above

**Review by jonmattgray — APPROVED** [review]

File: .github/workflows/test.yml (hunk 3 above)
Comment (dc-andysign) [reply .github/workflows/test.yml]: 👍

File: .github/workflows/test.yml (hunk 2 above)
Comment (dc-andysign) [reply .github/workflows/test.yml]: 👍

File: .github/workflows/release.yml (hunk 1 above)
Comment (dc-andysign) [reply .github/workflows/release.yml]: 👍

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -29,29 +30,73 @@ jobs:
...
+        run: npm install # maybe this should be changed to npm ci
+      - name: Dependency Check
+        run: npm run depcheck
+
+  tests:
+    name: Run tests
     strategy:
       matrix:
-        node-version: [18.x]
+        test-type: [unit]
```
[the remark this answers is not in the document]
Comment (dc-andysign) [reply .github/workflows/release.yml]: 👍

**Pull request merged**